vpr/src/place/initial_placement.cpp

```diff
@@ -549,47 +547,188 @@ static std::vector<ClusterBlockId> find_centroid_loc(const t_pl_macro& pl_macro,
+    // closer it is, the faster the BFS will find the best solution.
+    t_physical_tile_loc src_grid_loc(src_flat_loc.x, src_flat_loc.y, src_flat_loc.layer);
+    const t_physical_tile_loc compressed_src_loc = compressed_block_grid.grid_loc_to_compressed_loc_approx(src_grid_loc);
+
+    // Weighted-BFS search the compressed grid for an empty compatible subtile.
+    size_t num_rows = compressed_block_grid.get_num_rows(0);
+    size_t num_cols = compressed_block_grid.get_num_columns(0);
+    vtr::NdMatrix<bool, 2> visited({num_cols, num_rows}, false);
+    float best_dist = std::numeric_limits<float>::max();
+    t_pl_loc best_loc(OPEN, OPEN, OPEN, OPEN);
+
+    std::queue<t_physical_tile_loc> loc_queue;
+    loc_queue.push(compressed_src_loc);
+    while (!loc_queue.empty()) {
+        // Pop the top element off the queue.
+        t_physical_tile_loc loc = loc_queue.front();
+        loc_queue.pop();
+
+        // If this location has already been visited, skip it.
+        if (visited[loc.x][loc.y])
+            continue;
+        visited[loc.x][loc.y] = true;
+
+        // Get the distance from this loc to the src_loc in grid space
+        // Note: In compressed space, distances are not what they appear. We are
+        //       using the true grid positions to get the truly closest loc.
+        //       Here, we use the flat centroid of the macro and measure its
+        //       distance from the center of this grid loc (hence the 0.5 offset).
+        auto grid_loc = compressed_block_grid.compressed_loc_to_grid_loc(loc);
+        float grid_dx = std::abs((float)grid_loc.x + 0.5f - src_flat_loc.x);
+        float grid_dy = std::abs((float)grid_loc.y + 0.5f - src_flat_loc.y);
+        float grid_dist = grid_dx + grid_dy;
+        // If this distance is worst than the best we have seen or is outside
+        // of the search distance, do not explore it or its neighbors.
+        // NOTE: This prune is always safe (i.e. it will never remove a better
+        //       solution) since this is a spatial graph and our objective is
+        //       positional distance. The un-visitied neighbors of a node should
+        //       have a higher distance than the current node.
+        if (grid_dist >= best_dist)
+            continue;
+
+        // In order to ensure our BFS finds the closest compatible location, we
+        // traverse compressed grid locations which may not actually be valid
+        // (i.e. no tile exists there). This is fine, we just need to check for
+        // them to ensure we never try to put a cluster there.
+        bool is_valid_compressed_loc = false;
+        const auto& compressed_col_blk_map = compressed_block_grid.get_column_block_map(loc.x, 0);
+        if (compressed_col_blk_map.count(loc.y) != 0)
+            is_valid_compressed_loc = true;
+
+        // If this distance is better than the best we have seen so far, try
+        // to see if this is a better solution.
+        if (is_valid_compressed_loc) {
```

Comment:
This is a fairly long routine. The logic is pretty clean, but I wonder if there's a helper function you can extract to shorten it? The code below for example could be a small helper function to check if there is a sub_tile at a location.